python/dgl/backend/mxnet/tensor.py

```diff
@@ -246,3 +257,124 @@ def zerocopy_to_numpy(arr):
+        grad_lhs = nd.empty((lhs_data_nd.shape[0],) + feat_shape,
+                            ctx=grad_out.context, dtype=grad_out.dtype)
+        K.backward_lhs_binary_op_reduce(
+            self.reducer, self.binary_op, self.graph, self.lhs, self.rhs,
+            lhs_data_nd, rhs_data_nd, out_data_nd, grad_out_nd,
+            zerocopy_to_dgl_ndarray_for_write(grad_lhs), self.lhs_map[1],
+            self.rhs_map[1], self.out_map[1])
+        grad_lhs = _reduce_grad(grad_lhs, lhs_data_nd.shape)
+        grad_rhs = nd.empty((rhs_data_nd.shape[0],) + feat_shape,
+                             ctx=grad_out.context, dtype=grad_out.dtype)
+        K.backward_rhs_binary_op_reduce(
+            self.reducer, self.binary_op, self.graph, self.lhs, self.rhs,
+            lhs_data_nd, rhs_data_nd, out_data_nd, grad_out_nd,
+            zerocopy_to_dgl_ndarray_for_write(grad_rhs), self.lhs_map[1],
+            self.rhs_map[1], self.out_map[1])
+        grad_rhs = _reduce_grad(grad_rhs, rhs_data_nd.shape)
+        return grad_lhs, grad_rhs
+
+
+def binary_reduce(reducer, binary_op, graph, lhs, rhs, lhs_data, rhs_data,
+                  out_size, lhs_map, rhs_map, out_map):
+    func = BinaryReduce(reducer, binary_op, graph, lhs, rhs, out_size, lhs_map,
+                        rhs_map, out_map)
+    return func(lhs_data, rhs_data)
+
+
+class CopyReduce(mx.autograd.Function):
+    def __init__(self, reducer, graph, target, out_size, in_map, out_map):
+        super(CopyReduce, self).__init__()
+        self.reducer = reducer
+        self.graph = graph
+        self.target = target
+        self.out_size = out_size
+        self.in_map = in_map
+        self.out_map = out_map
+
+    def forward(self, in_data):
+        feat_shape = in_data.shape[1:]
+        out_data = nd.empty((self.out_size,) + feat_shape,
+                            ctx=in_data.context, dtype=in_data.dtype)
+        in_data_nd = zerocopy_to_dgl_ndarray(in_data)
+        out_data_nd = zerocopy_to_dgl_ndarray_for_write(out_data)
+        K.copy_reduce(
+            self.reducer, self.graph, self.target, in_data_nd, out_data_nd,
+            self.in_map[0], self.out_map[0])
+        self.save_for_backward(in_data_nd, out_data_nd)
+        return out_data
+
+    def backward(self, grad_out):
+        in_data_nd, out_data_nd = self.saved_tensors
+        grad_out_nd = zerocopy_to_dgl_ndarray(grad_out)
+        grad_in = nd.empty(in_data_nd.shape, ctx=grad_out.context,
+                            dtype=grad_out.dtype)
+        K.backward_copy_reduce(
+            self.reducer, self.graph, self.target, in_data_nd, out_data_nd,
+            grad_out_nd, zerocopy_to_dgl_ndarray_for_write(grad_in),
+            self.in_map[1], self.out_map[1])
+        return grad_in
+
+
+def copy_reduce(reducer, graph, target, in_data, out_size, in_map, out_map):
+    func = CopyReduce(reducer, graph, target, out_size, in_map, out_map)
+    return func(in_data)
+
+
+def _reduce_grad(grad, shape):
```

Comment:
what is this function for?

Reply:
If the forward pass does broadcasting on certain feature dimension, then in the backward pass the gradient should be reduced on that dimension. Currently for simplicity, DGL C++ kernel does not do the gradient reduction, and therefore, it's done in python side.

Reply:
Please add comment

Reply:
Done